qrproofs/qnr.go

```diff
@@ -0,0 +1,219 @@
+package qrproofs
+
+import (
+	"errors"
+	"github.com/xlab-si/emmy/common"
+	"github.com/xlab-si/emmy/dlog"
+	"math/big"
+)
+
+// Statistical zero-knowledge proof	of quadratic non-residousity (implemented for historical
+// reasons).
+
+func QNR(y *big.Int, qr *dlog.QR) (bool, error) {
+	prover := NewQNRProver(qr, y)
+	verifier := NewQNRVerifier(qr, y)
+	m := qr.N.BitLen()
+
+	for i := 0; i < m; i++ {
+		w, pairs := verifier.GetChallenge()
+		prover.SetProofRandomData(w)
+		// get challenge from prover for proving that verifier is not cheating
+		randVector := prover.GetChallenge()
+
+		verProof := verifier.GetProofData(randVector)
+
+		verifierIsHonest := prover.Verify(pairs, verProof)
+		if !verifierIsHonest {
+			err := errors.New("verifier is not honest")
+			return false, err
+		}
+
+		typ, err := prover.GetProofData(w)
+		if err != nil {
+			return false, nil
+		}
+
+		proved := verifier.Verify(typ)
+		if proved == false {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+type QNRProver struct {
+	QR *dlog.QR
+	Y  *big.Int
+	w  *big.Int
+}
+
+func NewQNRProver(qr *dlog.QR, y *big.Int) *QNRProver {
+	prover := QNRProver{
+		QR: qr,
+		Y:  y,
+	}
+
+	return &prover
+}
+
+func (prover *QNRProver) GetChallenge() []int {
+	m := prover.QR.N.BitLen()
+	var randVector []int
+	for i := 0; i < m; i++ {
+		// todo: remove big.Int
+		b := common.GetRandomInt(big.NewInt(2)) // 0 or 1
+		var r int
+		if b.Cmp(big.NewInt(0)) == 0 {
+			r = 0
+		} else {
+			r = 1
+		}
+		randVector = append(randVector, r)
+	}
+	return randVector
+}
+
+func (prover *QNRProver) SetProofRandomData(w *big.Int) {
+	prover.w = w
+}
+
+func (prover *QNRProver) GetProofData(challenge *big.Int) (int, error) {
+	isQR, err := prover.QR.IsQR(challenge)
+	if err != nil {
+		return 0, err
+	}
+	var typ int
+	if isQR {
+		typ = 1 // challenge is of type r^2
+	} else {
+		typ = 2 // challenge is of type r^2 * y
+	}
+	return typ, nil
+}
+
+func (prover *QNRProver) Verify(pairs, verProof []*common.Pair) bool {
+	for ind, proofPair := range verProof {
+		if proofPair.B.Cmp(big.NewInt(0)) == 0 {
+			t := prover.QR.Multiply(proofPair.A, proofPair.A)
+			Aw := prover.QR.Multiply(pairs[ind].A, prover.w)
+			Bw := prover.QR.Multiply(pairs[ind].B, prover.w)
+
+			if (t.Cmp(Aw) != 0) && (t.Cmp(Bw) != 0) {
+				return false
+			}
+		} else {
+			r1Squared := prover.QR.Multiply(proofPair.A, proofPair.A)
+			r2Squared := prover.QR.Multiply(proofPair.B, proofPair.B)
+			r2Squaredy := prover.QR.Multiply(r2Squared, prover.Y)
+			pair := pairs[ind]
+
+			if !((r1Squared.Cmp(pair.A) == 0 && r2Squaredy.Cmp(pair.B) == 0) ||
+				(r1Squared.Cmp(pair.B) == 0 && r2Squaredy.Cmp(pair.A) == 0)) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+type QNRVerifier struct {
+	QR    *dlog.QR
+	x     *big.Int
+	y     *big.Int
+	typ   int
```

Comment:
I would prefer to rename this field to `verifierType` or `qrType` or something else that makes sense. I know `type` is reserved, but just `typ` kind of bothers me, sorry :smile: .
The same goes for all the other usages of `typ`.